### PR TITLE
Runtime config 

### LINKED ISSUE #167: Argument for `SDL_Delay()` underflows in context.hpp
### Problem

Argument for `SDL_Delay()` in `Context::Update()` underflows when the `FPS_CAP` is low enough to make `frameTime - Util::Time::GetDeltaTime() < 0` 

### Steps to reproduce

- Have a low `FPS_CAP` in `config.hpp`
- Have a high delta time

### Expected behavior

Not to delay for a very long time like 4294967295 ms.

### PTSD version

c3642c5171939378e07762bef79841c79dc665c7

### Operating system/version

Windows 11 Pro

### CMake version

3.27.8

### CMakeCache.txt

[CMakeCache.txt](https://github.com/ntut-open-source-club/practical-tools-for-simple-design/files/14848809/CMakeCache.txt)

### OpenGL Info

```txt
[info] OpenGL Info
 [info]   Vendor: NVIDIA Corporation
 [info]   Renderer: NVIDIA GeForce RTX 3070/PCIe/SSE2
 [info]   Version: 4.1.0 NVIDIA 551.23
 [info]   GLSL Version: 4.10 NVIDIA via Cg compiler
```

### PR DESCRIPTION
**Pull request by Dragonfly911117**
Implement #153
## How to Use
To use, place a `config.json` file either in the current directory (`.`) or the parent directory (`..`). If both directories contain the file, the one in the current directory will be read.

Example directory structure:
```
├── build/
│   ├── YOUR_GAME
│   └── config.json <- place here
└── config.json <- or here
```

The `config.json` file can contain any number of configurations, each in the following format:
```
{
  "title": "<string, default is 'Practice-Tools-for-Simple-Design'>",
  "window_pos_x": <int, default is SDL_WINDOWPOS_UNDEFINED (wherever the OS places the window)>,
  "window_pos_y": <int, default is SDL_WINDOWPOS_UNDEFINED (wherever the OS places the window)>,
  "window_width": <int, default is 1280>,
  "window_height": <int, default is 720>,
  "default_log_level": <int in range [0, 5], default is 2 (Util::Logger::Level::INFO)>,
  "fps_cap": <int, default is 60>
}
```

If a key is not present in the `config.json`, the default value will be used.

### CODE REVIEW

File: include/Util/Config.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+template <typename T>
+T Config::GetAs(const std::string &key) {
+    auto value = Config::Get(key);
+    if (value.is_null()) {
+        return T{};
+    }
+    return value.get<T>();
+}
```
Comment (ntut-rick) [inline include/Util/Config.hpp:98]: null value and default value is indistinguishable

File: src/config.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#include "config.hpp"
+#include "SDL_image.h"
+
+#include <fstream>
+
+std::string PTSD_Config::TITLE = "Practice-Tools-for-Simple-Design";
```
Comment (onon1101) [inline src/config.cpp:6]:
```suggestion
std::string_view PTSD_Config::TITLE = "Practice-Tools-for-Simple-Design";
```

File: include/config.hpp [hunk 3]
```diff
@@ -1,25 +1,35 @@
...
 #include "pch.hpp" // IWYU pragma: export
 
-#include "Util/Logger.hpp"
+namespace Core {
+class Context;
+}
+
+struct PTSD_Config {
+public:
+    static std::string TITLE;
```
Comment (onon1101) [inline include/config.hpp:13]:
```suggestion
    static std::string_view TITLE;
```

File: src/config.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        file = std::ifstream("config.json");
+    } else if (std::filesystem::exists("../config.json")) {
+        file = std::ifstream("../config.json");
+    } else {
+        LOG_WARN("config.json not found, using default configurations.");
+        return;
+    }
+    file >> j;
+    file.close();
+    TITLE = j["title"].get<std::string>();
```
Comment (onon1101) [inline src/config.cpp:31]:
```suggestion
    TITLE = j["title"].get<std::string_view>();
```

**Review by onon1101 — COMMENTED** [review]
不曉得是不是能從 `std::string` 改成 c17 的 `std::string_view`, 可能會對記憶體分配好一點？

**Comment by onon1101** [comment]
並且在其他的地方以 `data()` 等方式存取, (具體可能要看 cpp 的文件)

File: src/config.cpp (hunk 2 above)
Comment (Dragonfly911117) [reply src/config.cpp]: Suggestion applied.

File: include/config.hpp
Comment (ntut-rick) [inline include/config.hpp:1]: IMO static or singleton pattern is not necessary, just use normal class and provide a default global variable is fine.

**Review by ntut-rick — CHANGES_REQUESTED** [review]

**Review by NOOBDY — APPROVED** [review]

**Review by ntut-rick — APPROVED** [review]

**Pull request merged**